### PR TITLE
Android X (P)

### PR DESCRIPTION
**Pull request by hannesa2**
I was forced to use Android P as well, so it's the move to P as well
 
-----------

Edited by @davigonz

About **Android P**, I'm gathering some changes from [here](https://developer.android.com/about/versions/pie/android-9.0-changes-28) that might affect the ownCloud Android app when targeting Android P:

- **Network TLS enabled by default**: if your app targets Android 9 or higher, the isCleartextTrafficPermitted() method returns false by default. If your app needs to enable cleartext for specific domains, you must explicitly set cleartextTrafficPermitted to true for those domains in your app's Network Security Configuration. Reproducible and fix in https://github.com/owncloud/android/pull/2402#issuecomment-456881009

- **Foreground services:**  apps that target Android P or higher and use foreground services must request the FOREGROUND_SERVICE permission. E.g.: https://github.com/owncloud/android/blob/master/owncloudApp/src/main/java/com/owncloud/android/files/services/FileUploader.java#L306

CC @jesmrec @hannesa2 

___

BUGS & IMPROVEMENTS

- [x] (1) Camera uploads https://github.com/owncloud/android/pull/2402#issuecomment-460961991

### CODE REVIEW

**Comment by hannesa2** [comment]
Ok, there are some `-alpha02` included, either 

- we wait till they are non alpha or
- we go with `-alpha02` (my company uses it in production without any issues , 5.000.000+ installations) or
- we uses older stable versions, not sure if it work for all libs

**Comment by davigonz** [comment]
Thanks a lot for this contribution @hannesa2 , we'd rather not use alpha versions, would it be possible to keep using stable versions just in the alpha ones?

**Comment by hannesa2** [comment]
I reverted to stable libs. 
But I'm not convinced, because it's just number cosmetic. After 2.0.0 a 2.0.1 will come too. On the other side, using old ones means to waiver some improvement in beta

File: owncloudApp/build.gradle [hunk 1]
```diff
@@ -3,48 +3,43 @@ apply plugin: 'com.android.application'
...
-    implementation "com.android.support:support-v4:${supportLibraryVersion}"
-    implementation "com.android.support:design:${supportLibraryVersion}"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation "androidx.legacy:legacy-support-v4:$androidX"
+    implementation 'com.google.android.material:material:1.0.0'
     implementation 'com.jakewharton:disklrucache:2.0.2'
     implementation 'com.google.android.exoplayer:exoplayer:r2.2.0'
     implementation 'com.andrognito.patternlockview:patternlockview:1.0.0'
-    implementation "com.android.support:appcompat-v7:${supportLibraryVersion}"
+    implementation 'androidx.appcompat:appcompat:1.0.0'
```
Comment (davigonz) [inline owncloudApp/build.gradle]: Why don't you get the version from `$androidX` here?

**Comment by davigonz** [comment]
Apart from the requested changes above, I have compiled the code, install the app and I get the next error when checking the server in login view, which appears as `Unknown error` in the app, any idea?

```
2019-01-23 17:53:27.454 25087-25119/com.owncloud.android.debug E/GetRemoteStatusOperation: Connection check at http://192.168.0.173:10000: Unrecovered transport exception
    java.net.UnknownServiceException: CLEARTEXT communication to 192.168.0.173 not permitted by network security policy
        at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:147)
        at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
        at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
        at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
```

**Comment by davigonz** [comment]
About Android P, I'm gathering here some changes that might affect the ownCloud Android app when targeting Android P:

- **Network TLS enabled by default**: if your app targets Android 9 or higher, the isCleartextTrafficPermitted() method returns false by default. If your app needs to enable cleartext for specific domains, you must explicitly set cleartextTrafficPermitted to true for those domains in your app's Network Security Configuration.

- In res/xml/network_security_config.xml:

```xml
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <base-config cleartextTrafficPermitted="true" />
</network-security-config>
```

- In AndroidManifest.xml

```xml
android:networkSecurityConfig="@xml/network_security_config"
```

- **Foreground services:**  Apps that target Android P or higher and use foreground services must request the FOREGROUND_SERVICE permission. E.g.: https://github.com/owncloud/android/blob/master/owncloudApp/src/main/java/com/owncloud/android/files/services/FileUploader.java#L306

**Comment by michaelstingl** [comment]
> But a move in the other direction could be an introduce of (dynamic) cert pinning which prevents from a man-in-the-middle attacks -just as an idea-

👍 

@hannesa2 you mean like in https://github.com/owncloud/android/issues/934 ? Or better open another issue to discuss details?

**Comment by hannesa2** [comment]
Exactly https://github.com/owncloud/android/issues/934 ! (but I it's almost 3 years just an idea)
I will not promise anything, but maybe I can deliver something, when I generalize similar somewhere else 

And one more additional step could be a 2 phase authentication, like Google does when a new IP/device tries to connect with a mail notification/confirmation. But with this, the backend needs some work. But to be honest, this needs too much live/spare time from my side.

When it comes to my focus I can open a ticket for sure, but currently my main interest is the non working (already closed) upload https://github.com/owncloud/android/issues/2394 
Btw you can open a ticket too, if you see a sense to have an additional ticket

File: owncloudApp/build.gradle (hunk 1 above)
Comment (hannesa2) [reply owncloudApp/build.gradle]: because 
`androidX` is `1.0.0` and 
`androidx.appcompat:appcompat` has currently `1.0.2`
But my fault I forgot to apply `1.0.2` here

done

**Comment by hannesa2** [comment]
What should be done here ?
Allow 'http' and be more un-secure or follow the Goggle suggestion and only allow `https` and do nothing ?
@davigonz @jesmrec @michaelstingl

**Comment by davigonz** [comment]
> What should be done here ?
Allow http and be more un-secure or follow the Goggle suggestion and only allow https and do nothing ?

I would not get rid of http IMHO, is insecure for sure but ownCloud is a self-hosted solution and therefore, each end user/company should be the responsible for the security of his ownCloud instance. 

Disallowing http will likely trigger complains since we are making decisions that do not belong to us.

> But a move in the other direction could be an introduce of (dynamic) cert pinning which prevents from a man-in-the-middle attacks -just as an idea-

About this, I think that we are already checking the SSL certificate against a stored one, as is suggested in #934. Have a look at https://github.com/owncloud/android-library/blob/940f0b9bb042e3c318e4cbeff71782e0d8733f93/owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/HttpClient.java#L78.

**Comment by davigonz** [comment]
@hannesa2 can you please apply the changes about [http](https://github.com/owncloud/android/pull/2402#issuecomment-456881009) and [foreground services ](https://github.com/owncloud/android/pull/2402#issue-239732799) so we can push this to be included in 2.10.0? Thanks

**Comment by hannesa2** [comment]
[http](https://github.com/owncloud/android/pull/2402#issuecomment-456881009) :
I followed https://developer.android.com/training/articles/security-config and applied the default configuration for apps targeting Android 6.0 (API level 23) and lower

**Comment by hannesa2** [comment]
Foreground Service:

I added the missing ones, but to be honest I did only code changes and didn't test it.

In following files, in my point of view, there are too much conditions to run into `startForeground()` but you already have it on master, so I didn't changed it.
This would be ok
`Build.VERSION.SDK_INT >= Build.VERSION_CODES.O`  but the other ones ?

FileDownloader.java
`if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && (isAvailableOfflineFile || retryDownload))`

FileUploader.java
`if ((createdBy == CREATED_AS_CAMERA_UPLOAD_PICTURE || createdBy == CREATED_AS_CAMERA_UPLOAD_VIDEO ||
                isAvailableOfflineFile || isRequestedFromWifiBackEvent) &&
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {`

**Comment by davigonz** [comment]
> FileDownloader.java
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && (isAvailableOfflineFile || retryDownload))

You have a comment just below that line that explains why do we need those checks. We create the service to check available offline files and retry downloads by using `startForegroundService` and we have to call `startForeground`  within five seconds only for those cases.

https://github.com/owncloud/android/blob/2b15b5bab69d83e22f7ccb70b6868eb49edd3341/owncloudApp/src/main/java/com/owncloud/android/files/services/FileDownloader.java#L193-L197

> if ((createdBy == CREATED_AS_CAMERA_UPLOAD_PICTURE || createdBy == CREATED_AS_CAMERA_UPLOAD_VIDEO || isAvailableOfflineFile || isRequestedFromWifiBackEvent) && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

Have a look at the comment below: https://github.com/owncloud/android/blob/2b15b5bab69d83e22f7ccb70b6868eb49edd3341/owncloudApp/src/main/java/com/owncloud/android/files/services/FileUploader.java#L300-L304

**Comment by davigonz** [comment]
@hannesa2 regarding the foreground services, I think you should not need to touch any code, as you did in https://github.com/owncloud/android/pull/2402/commits/1d843409951e2b636b2f6c226d88d80ec1f4aebe. Including [FOREGROUND_SERVICE](https://developer.android.com/reference/android/Manifest.permission#FOREGROUND_SERVICE) permission in AndroidManifest would be enough, as the [Android P documentation](https://developer.android.com/about/versions/pie/android-9.0-changes-28) says.

**Comment by hannesa2** [comment]
- I reverted 1d84340
- I added FOREGROUND_SERVICE 
- I removed one lint warning
image

**Comment by davigonz** [comment]
@hannesa2 ok thanks

File: owncloudApp/build.gradle [hunk 2]
```diff
@@ -60,19 +55,19 @@ tasks.withType(Test) {
 }
 
 android {
-    compileSdkVersion 26
+    compileSdkVersion 28
 
     defaultConfig {
-        minSdkVersion 14
-        targetSdkVersion 26
+        minSdkVersion 15
```
Comment (davigonz) [inline owncloudApp/build.gradle]: Let's keep this in 14, I know is a pretty old version but we will upgrade it after discussing it in https://github.com/owncloud/android/issues/2431

**Comment by davigonz** [comment]
@hannesa2 just a minor change and we can move it forward, could you make the Android X and P upgrade in the library as well? Thanks

File: owncloudApp/build.gradle (hunk 2 above)
Comment (hannesa2) [reply owncloudApp/build.gradle]: Google removed 14 from any statistic, because is as rare like a unicorn.
It's even not worth to discuss to revert to 14, you will loose 0,0001 % of users
It tried to revert
image

Let' go 15. Even 15 is like a unicorn

**Comment by davigonz** [comment]
Ready to test @jesmrec , though the library doesn't have Android X(P) yet, I suggested @hannesa2 to do it

**Comment by hannesa2** [comment]
I guess you mean this https://github.com/owncloud/android-library/pull/231.

**Comment by davigonz** [comment]
> I guess you mean this [owncloud/android-library#231](https://github.com/owncloud/android-library/pull/231).

Yes, thanks

**Comment by jesmrec** [comment]
Let's start a QA round

**Comment by jesmrec** [comment]
### (1)

Camera uploads are not working properly. Check these steps out:

1. Go to `Settings` and enable camera uploads for videos and pic
2. Minimize the app, open the camera and take a pic and a video
3. Wait 15 minutes, the app is not in foreground

Current: Items are not uploaded. They are in the uploads view with `App terminated`. The app finally leads to ANR:

>2019-02-06 10:45:45.085 1174-1206/? E/ActivityManager: ANR in com.owncloud.android.debug
    PID: 20006
    Reason: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{d7a9b0e u0 com.owncloud.android.debug/com.owncloud.android.files.services.FileUploader}
    Load: 4.4 / 4.23 / 4.29`

Expected: Items correctly uploaded after max. 15 minutes.

Google Pixel 2 v9 API 28

**Comment by hannesa2** [comment]
It seems https://github.com/owncloud/android/pull/2402#issuecomment-459262343 is not true.
I will reintroduce  1d84340 again, @jesmrec please have a look on it.

**Comment by hannesa2** [comment]
Interesting ! Maybe this is the reason for https://github.com/owncloud/android/issues/2394 ?
But unfortunately I can't build now. It seams google maven has an issue now, because 3h before PR was able to build. I will try it tomorrow

**Comment by jesmrec** [comment]
@hannesa2 with the latest commit the app crashes from scratch (Android P)

> Process: com.owncloud.android.debug, PID: 10795  android.app.RemoteServiceException: Bad notification for startForeground: java.lang.RuntimeException: invalid channel for service notification: Notification(channel=null pri=0 contentView=null vibrate=null sound=null defaults=0x0 flags=0x40 color=0xff1d2d44 vis=PRIVATE)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1738)

**Comment by hannesa2** [comment]
Do you have to complete stacktrace ?
I run my personal build https://github.com/hannesa2/owncloud-android/releases and don't see this (till now)

File: owncloudApp/build.gradle [hunk 3]
```diff
@@ -3,48 +3,44 @@ apply plugin: 'com.android.application'
...
-}
-
 dependencies {
     /// dependencies for app building
     implementation project(':owncloud-android-library:owncloudComLibrary')
-    implementation "com.android.support:support-v4:${supportLibraryVersion}"
-    implementation "com.android.support:design:${supportLibraryVersion}"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation "androidx.legacy:legacy-support-v4:$androidX"
+    implementation "androidx.annotation:annotation:1.0.1"
```
Comment (hannesa2) [inline owncloudApp/build.gradle]: without this line, build "release" run into an issue. That's why build in "debug" was too less

**Comment by hannesa2** [comment]
btw, I made a test picture and since 45 minutes nothing is transferred to my Owncloud.
You can close https://github.com/owncloud/android/issues/2394 but this doesn't help to fix it.

**Comment by hannesa2** [comment]
Now it syncs new pictures !

**Comment by davigonz** [comment]
> Camera uploads are not working properly. Check these steps out:
> 1. Go to `Settings` and enable camera uploads for videos and pic
> 2. Minimize the app, open the camera and take a pic and a video
> 3. Wait 15 minutes, the app is not in foreground

> Current: Items are not uploaded. They are in the uploads view with `App terminated`. The app finally leads to ANR:
> 2019-02-06 10:45:45.085 1174-1206/? E/ActivityManager: ANR in com.owncloud.android.debug
> PID: 20006
> Reason: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{d7a9b0e u0 com.owncloud.android.debug/com.owncloud.android.files.services.FileUploader}
> Load: 4.4 / 4.23 / 4.29`

I think this was not tested with the proper commit. It should've been tested with 406494134e35e13c2e1e6b80baf5e0842257a804 but I suspect it was tested with the commit before that (6ed4d309bd0c854780f4707ff46708e6beb423e4) cause I'm getting the error `Process: Bad notification for startForeground: java.lang.RuntimeException: invalid channel for service notification:`. 

Can you check that @jesmrec ?

> It seems [#2402 (comment)](https://github.com/owncloud/android/pull/2402#issuecomment-459262343) is not true.
I will reintroduce [1d84340](https://github.com/owncloud/android/commit/1d843409951e2b636b2f6c226d88d80ec1f4aebe) again, @jesmrec please have a look on it.

@hannesa2 Foreground services always need a notification and it does not make any sense to start every service in foreground as you did in https://github.com/owncloud/android/pull/2402/commits/e07901b9b2dc0f3db641a2dbad98cb119eb6a2f8, they make sense when uploading, downloading... 

The main problem here is related to the notifications. In Android P and after using FOREGROUND_SERVICE permission it seems we can not build notifications without parameters as we did in https://github.com/owncloud/android/blob/master/owncloudApp/src/main/java/com/owncloud/android/files/services/FileUploader.java#L306 and that was causing the error `Process: Bad notification for startForeground: java.lang.RuntimeException: invalid channel for service notification:`.

To sum up, we do not need https://github.com/owncloud/android/pull/2402/commits/e07901b9b2dc0f3db641a2dbad98cb119eb6a2f8 , using FOREGROUND_SERVICE in manifest and 72a4297c14c1d30780c4902a18045935bfd9e75e would be enough.

**Comment by hannesa2** [comment]
> To sum up, we do not need [e07901b](https://github.com/owncloud/android/commit/e07901b9b2dc0f3db641a2dbad98cb119eb6a2f8) , using FOREGROUND_SERVICE in manifest and [72a4297](https://github.com/owncloud/android/commit/72a4297c14c1d30780c4902a18045935bfd9e75e) would be enough.

I agree, [72a4297](https://github.com/owncloud/android/commit/72a4297c14c1d30780c4902a18045935bfd9e75e) fixes now the [e07901b](https://github.com/owncloud/android/commit/e07901b9b2dc0f3db641a2dbad98cb119eb6a2f8).
I can squash  (together with introduce 6dd0a4cc and revert d84f0bf02) it on request to summarize necessary changes, but in review you see summarize too 

Btw, `master` was not working for me as well, now with this PR it's working

**Comment by hannesa2** [comment]
> Can you check that @jesmrec ?

My opinion is, it's just necessary to test latest commit

**Comment by davigonz** [comment]
> I agree, [72a4297](https://github.com/owncloud/android/commit/72a4297c14c1d30780c4902a18045935bfd9e75e) fixes now the [e07901b](https://github.com/owncloud/android/commit/e07901b9b2dc0f3db641a2dbad98cb119eb6a2f8).
I can squash (together with introduce [6dd0a4c](https://github.com/owncloud/android/commit/6dd0a4ccc16fc576da3c3a6b3baf1c1ef925d026) and revert [d84f0bf](https://github.com/owncloud/android/commit/d84f0bf024c664076652c7a09c557d1e53e1fa18)) it on request to summarize necessary changes, but in review you see summarize too

@hannesa2 the `startForegroundService` stuff is still in this PR, can you have a look at it?

**Comment by hannesa2** [comment]
please test it first. Without it, we've seen a crash https://github.com/owncloud/android/pull/2402#issuecomment-461133730

**Comment by davigonz** [comment]
> please test it first. Without it, we've seen a crash [#2402 (comment)](https://github.com/owncloud/android/pull/2402#issuecomment-461133730)

As I explained in https://github.com/owncloud/android/pull/2402#issuecomment-462302042, I tested without it and I'm not getting that crash but the notification error. 

@jesmrec can you please test af756ba283513ab4a3c9a94268ebde08b9039e34 and confirm there's no crash there? If so, we can remove the `startForegroundService` stuff

**Comment by jesmrec** [comment]
With commit 406494134e35e13c2e1e6b80baf5e0842257a804:

 - Installation: OK
 - URL check and authentication:  OK
 - Camera uploads: Fails with https://github.com/owncloud/android/pull/2402#issuecomment-460961991

With commit 6ed4d30

 - Installation: OK
 - URL check and authentication:  OK
 - Camera uploads: Fails. In this case, pics are not either enqueued (in the previous one, they were). I stumbled upon this trace:

> 2019-02-12 11:01:29.270 22838-22838/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.owncloud.android.debug, PID: 22838
    java.lang.RuntimeException: Unable to start service com.owncloud.android.files.services.FileUploader@a120a3d with Intent { cmp=com.owncloud.android.debug/com.owncloud.android.files.services.FileUploader (has extras) }: java.lang.SecurityException: Permission Denial: startForeground from pid=22838, uid=10482 requires android.permission.FOREGROUND_SERVICE
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3687)


with commit 72a4297c14c1d30780c4902a18045935bfd9e75e  (the latest one)

 - Installation: OK
 - URL check and authentication:  OK
 - Camera uploads: Works fine!!!
 - Additional check: av, offline files: OK

so, good news. In any case, a deeper testing should be performed to assure all scenarios are OK, but this is out of the scope of the current one.

**Comment by jesmrec** [comment]
@hannesa2 i have rechecked the whole feature by dropping the commit c2db38e849b9fbdf54fea4e57f055c46ddfab682, and it works properly. No crashes in authentication and uploads run correctly (camera uploads and manual uploads). 

Maybe we can get rid of such commit, so it does not introduce anything sensible for the described feature, and will make a cleaner history. 

Sorry for the noise.

**Comment by hannesa2** [comment]
ok, I removed the 
* introduce
* revert
* re-intoduce

commit, and kept all except my fix. (To be honest, because it worked for me, I simply followed your suggestion and didn't tested it)
@jesmrec please take over

**Comment by jesmrec** [comment]
After a new check:

- Services job-based (camera uploads and av. offline syncing) are working properly in Android 9
- Same checks works fine in Android 8 and 7. 

We can move forward, and new checks will be performed in the following release stage.

Approved

**Comment by jesmrec** [comment]
Please, update the branch for the final merge

**Comment by hannesa2** [comment]
Sounds good ! Rebase is done
